### ISSUE TITLE
Preparations and considerations for SD3 [Discussion Thread] 

### ISSUE THREAD

**Issue by silveroxides** [klimaleksus/stable-diffusion-webui-embedding-merge#19]
Seems SD3 will handle some stuff a bit different. ComfyUI has already added preliminary support for it and it might be something worth looking into. Especially the [timestep embedding](https://github.com/comfyanonymous/ComfyUI/blob/8c4a9befa7261b6fc78407ace90a57d21bfe631e/comfy/ldm/modules/diffusionmodules/mmdit.py#L215) part. That is something new right?

[Link to the SD3 support commit](https://github.com/comfyanonymous/ComfyUI/commit/8c4a9befa7261b6fc78407ace90a57d21bfe631e)

**Comment by aleksusklim** [comment]
I think I'll rather wait for the full WebUI support and working local SD3 models, because that definitely would be buggy at first.

**Comment by ash-bluepollution95** [comment]
WebUI has full support as of now, the version i've been running has it.
SD 3.5 might be your better option due to liscencing

**Comment by aleksusklim** [comment]
Since Flux's release, I use only ComfyUI. And only with Flux, since it is strictly superior comparing to any other model. When I need SDXL (for in/outpaint), I go to Fooocus. Not a single time I have used SD1 since then.

I would rather somehow port EM to ComfyUI , at least for SD1/2 and SDXL.

To continue with SD3 otherwise, I need an article about how embeddings are working there internally, and a bunch of trained embedding files to start with.

**Comment by silveroxides** [comment]
@aleksusklim I also use primarily ComfyUI. Regarding both SD3 and Flux(if you set DualCLIPLoader to sd3), you can use CLIP G(SD3) and CLIP L(SD3 and FLux) embeddings and they actually work. When it comes to T5 it works a bit different. Overall ComfyUI acts a bit different with embeddings than A1111 did. For creation of embeddings for CLIP L and CLIP G, one would most likely have to port the logic from A1111. With T5 I have no Idea.

**Comment by ash-bluepollution95** [comment]
> Since Flux's release, I use only ComfyUI. And only with Flux, since it is strictly superior comparing to any other model. When I need SDXL (for in/outpaint), I go to Fooocus. Not a single time I have used SD1 since then.
> 
> I would rather somehow port EM to ComfyUI , at least for SD1/2 and SDXL.
> 
> To continue with SD3 otherwise, I need an article about how embeddings are working there internally, and a bunch of trained embedding files to start with.

I've been working on that already XD https://github.com/Ktiseos-Nyx/Comfyui_EmbeddingMerge_Node
BARELY THAT IS, like it's not really easy since i don't use ComfyUI natively.

**Comment by silveroxides** [comment]
@aleksusklim @klimaleksus I have created a custom node that at least allows for creation and saving of embedding in ComfyUI now and that support L, G and T5XXL. I'm going to give it a try to implement the parser and everything from this repo to see if I can reach some kind of semblance. 
[silveroxides/ComfyUI_EmbeddingToolkit](https://github.com/silveroxides/ComfyUI_EmbeddingToolkit)

**Comment by aleksusklim** [comment]
How embeddings are working in ComfyUI ?
Are they natively supported?

**Comment by silveroxides** [comment]
Yes. The ones created with this extension has worked fine. I just needed to figure out at which stage to fetch it before it was encoded to conds.

**Comment by silveroxides** [comment]
Oh and there is no limitations such as SDXL must have both L and G present. Each text encoder can be given their own embedding and if one has mutliple and some are not supported, then the supported ones in that embedding still gets loaded.

**Comment by aleksusklim** [comment]
Can you give a workflow that uses standard embedding mechanism for SDXL, where I could use arbitrary embeddings in prompts, just as I would use them in A1111? Or at least the best what can be done in Comfy?
(For the start, without any merging)

**Comment by silveroxides** [comment]
Sure. But the basic gist of it for adding embeddings in prompt is the syntax "embedding:path_to_embedding\embedding_filename" no need for tile extension and if it is in top folder in models/embeddings then you just do "embedding:embedding_filename".

This is as basic as they get. Put embeddings I used in the drive folder too. You can weight them too. Just do "(embedding:embedding_filename:1.1)".

https://drive.google.com/drive/folders/1eobHxcSgz5ZzbAG6uXNbeOPYyumdw13R?usp=sharing

**Comment by aleksusklim** [comment]
Thank you! Now I see how embeddings are used from UI side (didn't look at Comfy internals).

Are `embedding:*` patches cached? Meaning, if your node would be sitting in-between "CLIP" route (reading CLIP wire, having TEXT widget, outputting const string like `embedding:EM` wire that can be connected for CLIP Encode instead of normal prompt) – would ComfyUI read the same overwritten file again each time?

If they aren't cached, I think it would be possible to at least merge the whole prompt expression in one temporary embedding, letting Comfy use it instead of the whole prompt itself.

**Comment by silveroxides** [comment]
embeddings are only loaded upon execution of queue and that only the encoded conds are cached if no changes are done in the text input that the text encode node has or has connected. In essence, you actually have more control here than you would in A1111. Things are just not as gathered in one spot. The easiest way I have found to navigate the code is to use Github and just put text marker at a functions name and see where it is being used. I wrote an entire model architecture implementation this way that is now part of core ComfyUI. 
I think that if I attempt to go in to how part of it works in text here, I might only make it more confusing. 
When it comes to implementing EM in to ComfyUI the only large hurdle for me has been to dissect the EM code in order to remove A1111 code and locate what part of code that needs some form of input data and what that data is.

hmm. Oh one thing about embedding:*
It always reads the path indicated in prompt so if an EM parser would store a temp embedding, then the parser would only need to replace "EM parsed text portion" with the "embedding:*" representation with that path before forwarding to the internal prompt parser.

**Comment by aleksusklim** [comment]
So basically we can make a node that has CLIP input wire, TEXT widget/wire for prompt, and outputs TEXT wire for "EM prompt" that would have `EM_*` embedding names stuffed in place of merge expressions, for example `A photo of <'cat'+'dog'*0.8> at <'winter'+'summer'>` would output `A photo of embedding:EM/EM_1 at embedding:EM/EM_2` – to be connected with normal CLIP encoding.

This node must create temporary files with proper tensors stored as embeddings, just for the later CLIP node to read them by ComfyUI means.
(Plus, maybe some empty fake embedding named with a hash of the input prompt – to fight caching in cases when the user changes only EM weights without touching the main text)

If you already have a sample node that can render text/tokens to embeddings, the only major part left is porting the math parser from EM.
Is this what you need?

Essentially, you can try grabbing
https://github.com/klimaleksus/stable-diffusion-webui-embedding-merge/blob/ebe331a7bba00180e7c9b307b905646e980c853c/scripts/embedding_merge.py#L477-L745
replacing `clip` usage with whatever Comfy does to translate text into tensors. Note that my code references `clip` implicitly in other functions too, you'll have to adapt more code than just the main body.

My link points to the version before SDXL support, since the node processes just one CLIP instead of two simultaneously.

**Comment by Unarmedmann** [comment]
Sorry for off topic, but does Chroma and Qwen support embeddings? 

Could it be a way for people with low VRAM to avoid training LORAs and finetuning? I tried to ogle this, but found no answer. Seems like it might be really useful if it works.

**Comment by ash-bluepollution95** [comment]
Chroma and Qwen are Flux finetunes.  So no.
AFAIK: most DIT models don't require these -- I have NEVER seen a bit to be
able to even TRAIN a textual inversion for SD3.

**Comment by silveroxides** [comment]
@duskfallcrew
Ehm no not true. Chroma is not a Flux1 fine-tune. It is a pruned and pre-trained model with a de-distilled Schnell at its base.
Qwen is an entirely different architecture. What they have in common are that they are flow matching models.
Embedding as possible for both but it is trickier than it was with SDXL and SD1.5. 
Lastly, them being DiT models has nothing to do with their capabilities regarding embeddings.

**Comment by ash-bluepollution95** [comment]
Chroma is a Flux SCHNELL finetune :P
There. We're even.
Qwen somewhere i saw IT WAS a flux schnell base tho- THAT WAS on me if i'm
wrong though.

Well and sure i'll bite that i'm wrong on DIT, i'm always up for
re-schooling here :)
The thing is though, is have they released information on how to make them
--
Because on SDXL it's easy -- but SDXL's finetunes is another ballgame.

Illustrious proves trickier than Pony.
Noob ehh a bit easier then Illustrious but not by much
